src/Microsoft.OData.Client/ALinq/GroupByAnalyzer.cs

```diff
@@ -0,0 +1,529 @@
+            {
+                return false;
+            }
+
+            LambdaExpression resultSelector;
+            // Result selector should be a double argument lambda: (d2, d3) => ...
+            if (!ResourceBinder.PatternRules.MatchDoubleArgumentLambda(methodCallExpr.Arguments[2], out resultSelector))
+            {
+                return false;
+            }
+
+            // Analyze result selector - Must be a simple instantiation.
+            // A result selector where the body is a single argument lambda is also valid: (d2, d3) => d3.Average(d4 => d4.Amount)
+            // However, this presents a challenge since an alias is required in the aggregate transformation: aggregate(Amount with average as [alias])
+            // We could circumvent this by defaulting to [aggregate method]+[aggregate property], e.g., AverageAmount
+            // but this is deferred to a future release
+            if (resultSelector.Body.NodeType != ExpressionType.MemberInit && resultSelector.Body.NodeType != ExpressionType.New)
+            {
+                return false;
+            }
+
+            // Analyze key selector
+            if (!AnalyzeGroupByKeySelector(source, keySelector))
+            {
+                return false;
+            }
+
+            // Analyze result selector
+            GroupByResultSelectorAnalyzer.Analyze(source, resultSelector);
+
+            resourceExpr.Projection = new ProjectionQueryOptionExpression(resultSelector.Body.Type, resultSelector, new List<string>());
+
+            return true;
+        }
+
+        /// <summary>
+        /// Analyzes a GroupBy key selector for property/ies that the input sequence is grouped by.
+        /// </summary>
+        /// <param name="input">The input expression.</param>
+        /// <param name="keySelector">Key selector expression to analyze.</param>
+        /// <returns>true if analyzed successfully; false otherwise</returns>
+        private static bool AnalyzeGroupByKeySelector(QueryableResourceExpression input, LambdaExpression keySelector)
+        {
+            Debug.Assert(input != null, $"{nameof(input)} != null");
+            Debug.Assert(keySelector != null, $"{nameof(keySelector)} != null");
+
+            EnsureApplyInitialized(input);
+            Debug.Assert(input.Apply != null, $"{nameof(input)}.Apply != null");
+
+            // Scenario 1: GroupBy(d1 => [Constant])
+            ConstantExpression constExpr = keySelector.Body as ConstantExpression;
+            if (constExpr != null)
+            {
+                input.Apply.KeySelectorMap.Add(constExpr.Value.ToString(), constExpr);
+
+                return true;
+            }
+
+            // Scenario 2: GroupBy(d1 => d1.Property) and GroupBy(d1 = d1.NavProperty...Property)
+            if (keySelector.Body.NodeType == ExpressionType.MemberAccess)
+            {
+                MemberExpression memberExpr = (MemberExpression)keySelector.Body;

```

Comment:
nit: why do you cast ConstExpression and MemberExpression differently? is it possible to cast them similarly for uniformity?

Reply:
@ElizabethOkerio We treat them differently because they translate into quite different OData URLs.
For example,
```csharp
dataServiceContext.Sales.GroupBy(
    d1 => 1,
    (d2, d3) => new { AverageAmount = d3.Average(d4 => d4.Amount) })
```
where `1` is a `ConstExpression` is translated into
```
/Sales?$apply=aggregate(Amount with average as AverageAmount)
```

while

```csharp
dataServiceContext.Sales.GroupBy(
    d1 => d1.ProductId,
    (d2, d3) => new { AverageAmount = d3.Average(d4 => d4.Amount) })
````
where `d1.ProductId` is a `MemberExpression` is translated into:
```
/Sales?$apply=(groupby(ProductId),aggregate(Amount with average as AverageAmount))
```

Reply:
I think what I meant here is: using `keySelector.Body as ConstantExpression;` for const expressions and `(MemberExpression)keySelector.Body;` for MemberExpression.  but it's a minor detail. 

instead of: 
`keySelector.Body as ConstantExpression;` 
`keySelector.Body as MemberExpression`

if cast fails here then a nullreference exception is thrown

or 
`(ConstantExpression)keySelector.Body`
 `(MemberExpression)keySelector.Body`

if cast fails here an invalidcastexception is thrown. 

or these casts cannot fail?

Reply:
@ElizabethOkerio I made the improvement. Also moved the block for `MemberExpression` because I believe that scenario would be the most common